R/misc.R

```diff
@@ -79,6 +79,72 @@ model_printer <- function(x, ...) {
 is_missing_arg <- function(x)
   identical(x, quote(missing_arg()))
 
+# given a model object, return TRUE if:
+# * the model is supported without extensions
+# * the model needs an extension and it is loaded
+#
+# return FALSE if:
+# * the model needs an extension and it is _not_ loaded
+has_loaded_implementation <- function(spec_, engine_, mode_) {
+  if (isFALSE(mode_ %in% c("regression", "censored regression", "classification"))) {
+    mode_ <- c("regression", "censored regression", "classification")
+  }
+  eng_cond <- if (is.null(engine_)) {TRUE} else {quote(engine == engine_)}
+
+  avail <-
+    get_from_env(spec_) %>%
+    dplyr::filter(mode %in% mode_, !!eng_cond)
+  pars <-
+    utils::read.delim(system.file("models.tsv", package = "parsnip")) %>%
+    dplyr::filter(model == spec_, !!eng_cond, mode %in% mode_, is.na(pkg))
+
+  if (nrow(pars) > 0 || nrow(avail) > 0) {
+    return(TRUE)
+  }
+
+  FALSE
+}
+
+# construct a message informing the user that there are no
+# implementations for the current model spec / mode / engine.
+#
+# if there's a "pre-registered" extension supporting that setup,
+# nudge the user to install/load it.
+prompt_missing_implementation <- function(spec_, engine_, mode_) {
+  avail <-
+    show_engines(spec_) %>%
+    dplyr::filter(mode == mode_, engine == engine_)
+  all <-
+    utils::read.delim(system.file("models.tsv", package = "parsnip")) %>%
+    dplyr::filter(model == spec_, mode == mode_, engine == engine_, !is.na(pkg)) %>%
+    dplyr::select(-model)
+
+  if (identical(mode_, "unknown")) {
+    mode_ <- ""
+  }
+
+  msg <-
+    glue::glue(
+      "parsnip could not locate an implementation for `{spec_}` {mode_} model ",
+      "specifications using the `{engine_}` engine."
+    )
+
+  if (nrow(avail) == 0 && nrow(all) > 0) {
+    msg <-
+      glue::glue_collapse(c(
+        msg,
+        glue::glue_collapse(c(
+          "\nThe parsnip extension package {",
+          cli::col_yellow(all$pkg[[1]]),
+          "} implements support for this specification. \n",
+          "Please install (if needed) and load to continue.\n"
+        ))
```

Comment:
For consistence we use the [tidyverse style guide](https://style.tidyverse.org/documentation.html#r-code) which has no special font or anything for package names:

```suggestion
        glue::glue_collapse(c(
          "\nThis specification is available in the  ",
          cli::col_yellow(all$pkg[[1]]),
          " package. \n"
        ))
```

I'm a bit worried about the `\n` in this message and how that may interact with the rlang messages. Can we use the built-in bulleting, etc instead of hardcoding new lines into the message? 

Why is this `glue_collapse()` instead of just `glue()`. Because there can be more than one? Maybe we should use the pluralization from cli?

Reply:
I think this function should return a named character vector, with no name for the first element, `i` for the others, etc, to be used as `message` in `rlang::inform()`.

Reply:
Thank you for the nudges on making better use of glue and inform—learned a lot about those functions from your comments.

The message now reads:

``` r
library(parsnip)

C5_rules()
#> parsnip could not locate an implementation for `C5_rules` classification model specifications using the `C5.0` engine.
#> ℹ  The parsnip extension package rules implements support for this specification.
#> ℹ  Please install (if needed) and load to continue.
#> 
#> C5.0 Model Specification (classification)
#> 
#> Computational engine: C5.0

C5_rules(engine = "C5.0") %>% translate()
#> parsnip could not locate an implementation for `C5_rules` classification model specifications using the `C5.0` engine.
#> ℹ  The parsnip extension package rules implements support for this specification.
#> ℹ  Please install (if needed) and load to continue.
#> 
#> C5.0 Model Specification (classification)
#> 
#> Computational engine: C5.0
```

<sup>Created on 2022-05-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

and the code to get there is a good bit cleaner.

(Acknowledgement that I’m a serial violator of the no-special-package-name-formatting convention.🫠  Sorry.)